skcosmo/pcovr/pcovr.py

```diff
@@ -8,63 +9,233 @@
 from sklearn.decomposition._base import _BasePCA
 from sklearn.linear_model._base import LinearModel
 
-from .pcovr_distances import pcovr_covariance, pcovr_kernel
 from skcosmo.utils import eig_solver
 
 
-class PCovR(_BasePCA, LinearModel):
+def pcovr_covariance(mixing, X, Y, rcond=1e-12, return_isqrt=False):
     """
-    Performs Principal Covariates Regression, as described in `[S. de Jong and
-    H. A. L. Kiers, 1992] <https://doi.org/10.1016/0169-7439(92)80100-I>`_.
+    Creates the PCovR modified covariance
+
+    .. math::
+
+        \\mathbf{\\tilde{C}} = \\alpha \\mathbf{X}^T \\mathbf{X} +
+        (1 - \\alpha) \\left(\\left(\\mathbf{X}^T
+        \\mathbf{X}\\right)^{-\\frac{1}{2}} \\mathbf{X}^T
+        \\mathbf{\\hat{Y}}\\mathbf{\\hat{Y}}^T \\mathbf{X} \\left(\\mathbf{X}^T
+        \\mathbf{X}\\right)^{-\\frac{1}{2}}\\right)
+
+    where :math:`\\mathbf{\\hat{Y}}`` are the properties obtained by linear regression.
+
+    :param mixing: mixing parameter,
+                   as described in PCovR as :math:`{\\alpha}`, defaults to 1
+    :type mixing: float
+
+    :param X: Data matrix :math:`\\mathbf{X}`
+    :type X: array of shape (n x m)
+
+    :param Y: array to include in biased selection when mixing < 1
+    :type Y: array of shape (n x p)
+
+    :param rcond: threshold below which eigenvalues will be considered 0,
+                      defaults to 1E-12
+    :type rcond: float
+
+    """
+
+    C = np.zeros((X.shape[1], X.shape[1]), dtype=np.float64)
+
+    cov = X.T @ X
+
+    if mixing < 1 or return_isqrt:
+        # Do not try to approximate C_inv, it will affect results
+        C_inv = np.linalg.pinv(cov, rcond=rcond)
+        C_isqrt = np.real(scipy.linalg.sqrtm(C_inv))
+
+        # parentheses speed up calculation greatly
+        Y_hat = C_isqrt @ (X.T @ Y)
+        Y_hat = Y_hat.reshape((C.shape[0], -1))
+        Y_hat = np.real(Y_hat)
+
+        C += (1 - mixing) * Y_hat @ Y_hat.T
+
+    if mixing > 0:
+        C += (mixing) * cov
+
+    if return_isqrt:
+        return C, C_isqrt
+    else:
+        return C
+
+
+def pcovr_kernel(mixing, X, Y):
+    """
+    Creates the PCovR modified kernel distances
+
+    .. math::
+
+        \\mathbf{\\tilde{K}} = \\alpha \\mathbf{X} \\mathbf{X}^T +
+        (1 - \\alpha) \\mathbf{\\hat{Y}}\\mathbf{\\hat{Y}}^T
```

Comment:
FYI, you can use raw strings and not have to escape backslashes:
```suggestion
   r """
    Creates the PCovR modified kernel distances

    .. math::

        \mathbf{\tilde{K}} = \alpha \mathbf{X} \mathbf{X}^T +
        (1 - \alpha) \mathbf{\hat{Y}}\mathbf{\hat{Y}}^T
```

Reply:
A+ suggestion

Reply:
Okay, so suggestion is great in theory, but breaks black so I had to revert. Is there a reason for that?

Reply:
what's the error?

Reply:
No space between the r and the ", the r needs to be aligned with the lower quotations, and I _think_ you have to fix all "\\" in the quote, based upon what I had to change, in that order, to get it to work.

Reply:
So it looks like the raw string syntax do not work with triple quotes / multiline strings: https://stackoverflow.com/questions/20821431/how-is-it-possible-to-encode-triple-quotes-into-a-raw-string